client/request_test.go

```diff
@@ -0,0 +1,72 @@
+package client
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"testing"
+
+	"golang.org/x/net/context"
+)
+
+// TestSetHostHeader should set fake host for local communications, set real host
+// for normal communications.
+func TestSetHostHeader(t *testing.T) {
+	testURL := "/test"
+	testCases := []struct {
+		host            string
+		expectedHost    string
+		expectedURLHost string
+	}{
+		{
+			"unix:///var/run/docker.sock",
+			"",
+			"docker",
+		},
+		{
+			"npipe:////./pipe/docker_engine",
+			"",
+			"docker",
+		},
+		{
+			"tcp://0.0.0.0:4243",
+			"",
```

Comment:
I think the `test.expectedHost` is always `""`.
Now we just pass the apiPath (something with no schema like `/container/id/start`) into `http.NewRequest`. (See [here](https://github.com/docker/engine-api/blob/master/client/request.go#L132-L133))
While `http.NewRequest` will only set `request.Host` when the passed in url has Schema. (See [here](https://github.com/golang/go/blob/c1c7547f6ad7264c1d6eea3fc1645b2eab104694/src/net/http/request.go#L651), [here](https://github.com/golang/go/blob/c1c7547f6ad7264c1d6eea3fc1645b2eab104694/src/net/url/url.go#L417) and [here](https://github.com/golang/go/blob/c1c7547f6ad7264c1d6eea3fc1645b2eab104694/src/net/url/url.go#L483-L490).